app/Actions/Install/ApplyMigration.php

```diff
@@ -2,73 +2,97 @@
+	 * @param string[] $output list of messages
 	 *
-	 * @return array
+	 * @return void
 	 */
-	private function str_to_array($string, array &$output)
+	private function str_to_array(string $string, array &$output): void
 	{
 		$a = explode("\n", $string);
 		foreach ($a as $aa) {
 			if ($aa != '') {
 				$output[] = $aa;
 			}
 		}
-
-		return $output;
 	}
 
 	/**
-	 * @return bool
+	 * Runs the migration via the Artisan Facade.
+	 *
+	 * **TODO:** Consolidate with {@link \App\Actions\Update\Apply::migrate()}.
+	 *
+	 * **ATTENTION:** This method serves the same purpose as
+	 * `ApplyMigration::migrate()`.
+	 * The whole code around installation/upgrade/migration should
+	 * thoroughly be revised an refactored.
+	 *
+	 * @param string[] $output list of messages
+	 *
+	 * @return void
+	 *
+	 * @throws InstallationFailedException
 	 */
-	public function migrate(array &$output)
+	public function migrate(array &$output): void
 	{
 		Artisan::call('view:clear');
 		Artisan::call('migrate', ['--force' => true]);
 		$this->str_to_array(Artisan::output(), $output);
 
 		/*
-		 * We also double check there is no "QueryException" in the output (just to be sure).
+		 * We check there is no "QueryException" in the output (just to be sure).
 		 */
 		foreach ($output as $line) {
 			if (str_contains($line, 'QueryException')) {
-				// @codeCoverageIgnoreStart
-				return true;
-				// @codeCoverageIgnoreEnd
+				throw new InstallationFailedException('DB migration failed: ' . $line);
 			}
 		}
-
-		return false;
 	}
 
 	/**
-	 * @return bool
+	 * @param string[] $output list of messages
+	 *
+	 * @return void
+	 *
+	 * @throws InstallationFailedException
+	 * @throws FrameworkException
 	 */
-	public function keyGenerate(array &$output)
+	public function keyGenerate(array &$output): void
 	{
 		try {
 			Artisan::call('key:generate', ['--force' => true]);
 			$this->str_to_array(Artisan::output(), $output);
-			// @codeCoverageIgnoreStart
-		} catch (\Exception $e) {
-			$output[] = $e->getMessage();
-			$output[] = 'We could not generate the encryption key.';
+			if (!str_contains(end($output), 'Application key set successfully')) {
+				$output[] = 'We could not generate the encryption key.';
+				throw new InstallationFailedException('Could not generate encryption key');
+			}
 
-			return true;
+			// key is generated, we can safely remove that file (in theory)
+			$filename = base_path('.NO_SECURE_KEY');
+			if (file_exists($filename)) {
+				if (is_file($filename)) {
+					unlink($filename);
+				} else {
+					throw new InstallationFailedException('A filesystem object . ' . $filename . ' exists, but is not an ordinary file.');
+				}
+			}
+		} catch (\ErrorException $e) {
+			// possibly thrown by `unlink`
+			$output[] = $e->getMessage();
+			$output[] = 'Could not remove file `.NO_SECURE_KEY`.';
+			throw new InstallationFailedException('Could not remove file `.NO_SECURE_KEY`', $e);
```

Comment:
I am not sure if `$output[]` is needed as we are now throwing an exception...

Reply:
We do, because the parameter `$output` is passed by reference and thus becomes an output parameter which is used by the caller and returned to the web client even if an exception is thrown, because the caller catches the exception.

Haven't I said that I don't like this web installer/updater at all, have I? IMHO the whole thing should be replaced. Moreover, I don't like output parameters, because they tend to disguise the information flow. Your comment is prove of that.

However, I did not come up with this code. Improving this whole installer is a job for another PR.